### ISSUE TITLE
Forward Oridinal

### ISSUE THREAD

**Issue by PfiatDe** [hamflx/forward-dll#1]
Is it possible to also forward a function via the ordinal? At the moment, it seems like it is not possible to specify the ordinal in the function declaration.
Any plans to update this?
Thanks!

**Comment by hamflx** [comment]
I will do some research on how to do this.

**Comment by hamflx** [comment]
我提交了新的代码用以支持转发 ordinal，但是受限于我相关知识的缺乏所以支持的不是非常好。待我有时间我会讲一下我碰到的困难，也许你可以给我一些帮助。

建议你先克隆仓库，参考一下 examples/version 目录下的代码。注意，在支持 ordinal 转发之前，我提交了一些重构的代码（因此，你需要阅读一下 README.md 看一下新的使用方式，很简单），虽然以前的 forward_dll 宏仍然能用，但是要支持 ordinal 转发，需要以新的方式使用。

目前 ordinal 转发的实现方式需要两个步骤：

- #[forward(target = "...", ordinal)]，在 derive(ForwardModule) 的参数中，加入 ordinal，这一步会生成链接参数文件 `ordinal_link_args.txt`。
- 在你的 cdylib 的 crate 的 build.rs 中，读取该文件，并通过 cargo:rustc-link-arg= 将其打印出，作为链接器参数，参考 examples/version/build.rs。

重构后的版本已发布 `0.1.8`。

I've submitted new code to support forwarding ordinal, but it's not very good due to my lack of knowledge. When I have time I will talk about the difficulties I encountered, maybe you can give me some help.

I suggest you clone the repository first and refer to the code in the examples/version directory. Note that I committed some refactored code before supporting ordinal forwarding (so you need to read the README.md to see the new way of using it, it's easy), and while the old forward_dll macro still works, it needs to be used in a new way to support ordinal forwarding.

The current implementation of ordinal forwarding requires two steps:

- #[forward(target = "..." , ordinal)], add ordinal to the derive(ForwardModule) arguments, this step generates the link arguments file `ordinal_link_args.txt`.
- In your crate of cdylib's build.rs, read this file and print it out as linker arguments via cargo:rustc-link-arg=, refer to examples/version/build.rs.

the refactored version has been released `0.1.8`.

Translated with www.DeepL.com/Translator (free version)

**Comment by hamflx** [comment]
我尝试了如下的方案：

1. 直接增加 /export 链接参数，在 `build.rs` 加入如下代码：

   ```rust
   println!("cargo:rustc-link-arg=/export:GetFileVersionInfoA,@1");
   ```

   但是这没有任何作用，`GetFileVersionInfoA` 的 `ordinal` 仍然是 2。因为 `#[no_mangle]` 也隐含了将符号导出的意思。

2. 使用 `.def` 文件，这将导致 `crate` 原有的导出符号丢失，如 `DllMain`。

所以，最后我使用了一个较为麻烦的方法来实现这个功能：在导出的符号前增加一个前缀 `_`，如：`_GetFileVersionInfoA`，然后在链接时，传递 `/export:GetFileVersionInfoA=_GetFileVersionInfoA,@1` 给链接器，以此将导出的符号 `GetFileVersionInfoA` 指向 `_GetFileVersionInfoA`。

I tried the following approaches:

1. Adding the `/export` link parameter directly and inserting the following code in `build.rs`:

   ```rust
   println!("cargo:rustc-link-arg=/export:GetFileVersionInfoA,@1");
   ```

   However, this had no effect and the `ordinal` for `GetFileVersionInfoA` remained 2. This is because `#[no_mangle]` implies exporting the symbol.

2. Using a `.def` file would result in the loss of the original exported symbols of the crate, such as `DllMain`.

Therefore, I ended up using a more complicated approach to achieve this functionality: adding a prefix `_` to the exported symbol, such as `_GetFileVersionInfoA`, and then passing `/export:GetFileVersionInfoA=_GetFileVersionInfoA,@1` to the linker during linking to redirect the exported symbol `GetFileVersionInfoA` to `_GetFileVersionInfoA`.

**Comment by PfiatDe** [comment]
Crazy, that was fast. I will give it a try. Thanks a lot!

**Comment by hamflx** [comment]
我发布了 0.1.11，这个版本使用了 `Import Library` 和链接参数，实现了名称和 ordinal 的转发。这个版本可以看一下最新的文档 `Example 1`：<https://docs.rs/forward-dll/latest/forward_dll/>。

I have released 0.1.11, which uses `Import Library` and link parameters to implement name and ordinal forwarding. This version can be seen in the latest documentation `Example 1`: https://docs.rs/forward-dll/latest/forward_dll/.

Usage:

```toml
[build-dependencies]
forward-dll = "0.1.11"
```

`build.rs`:

```rust
use forward_dll::forward_dll;

fn main() {
  forward_dll("C:\\Windows\\System32\\version.dll").unwrap();
}
```

no `forward_dll` macro, no `#[derive(ForwardModule)]` macro.

**Comment by hamflx** [comment]
I have closed this issue, if you have any questions about this feature please feel free to reopen it or create a new issue.
